src/UCTSearch.cpp

```diff
@@ -45,16 +46,79 @@ UCTSearch::UCTSearch(GameState& g)
     m_root = std::make_unique<UCTNode>(FastBoard::PASS, 0.0f, 0.5f);
 }
 
+bool UCTSearch::advance_to_new_rootstate() {
+    if (!m_root || !m_last_rootstate) {
+        // No current state
+        return false;
+    }
+
+    if (m_rootstate.get_komi() != m_last_rootstate->get_komi()) {
+        return false;
+    }
+
+    auto depth =
+        (int) (m_rootstate.get_movenum() - m_last_rootstate->get_movenum());
+
+    if (depth < 0) {
+        return false;
+    }
+
+    auto test = std::make_unique<GameState>(m_rootstate);
+    for (auto i = 0; i < depth; i++) {
+        test->undo_move();
+    }
+
+    if (m_last_rootstate->board.get_ko_hash() != test->board.get_ko_hash()) {
+        // m_rootstate and m_last_rootstate don't match
+        return false;
+    }
+
+    // Try to replay moves advancing m_root
+    for (auto i = 0; i < depth; i++) {
+        test->forward_move();
+        const auto move = test->get_last_move();
+        m_root = m_root->find_child(move);
+        if (!m_root) {
+            // Tree hasn't been expanded this far
+            return false;
+        }
+        m_last_rootstate->play_move(move);
+    }
+
+    assert(m_rootstate.get_movenum() == m_last_rootstate->get_movenum());
+
+    if (m_last_rootstate->board.get_ko_hash() != test->board.get_ko_hash()) {
```

Comment:
You pointed out earlier this will hit even if there's a transposition, and that was a valid concern.

At the very least, it should use hash, not ko-hash, because the latter does not hash ko squares. (This is perhaps less confusing if one remembers that ko-hash is the hash used for detecting ko, not encoding it)

Reply:
I think it's not a problem in practice as it can only happens if you have to have multiple 4+ undos (which bots and self play don't have) and the consequence is tiny on the infrequent occasions it happens (eval for that position A will have been generated for position A with a different history).

Also it's not a regression from the current code (which also uses get_hash to verify positions are the same but doesn't check that all of m_lastroot moves match m_root moves).

If you still think it's a problem I suggest that we clear UCTSearch on "undo" or verify all move history between m_root and m_lastroot.